### PR TITLE
Messaging - As a User, I can contact my BC

### PR DESCRIPTION
**Pull request by victor-paumier**
- Let user contact their BCs (channel user / all BCs of the section)
- fix upload assets
- global messaging improvement

### CODE REVIEW

File: src/Controller/UserController.php [hunk 1]
```diff
@@ -158,13 +158,19 @@ public function showReport(User $user): Response
...
     ) {
-        $user = $this->getUser();
         $query = $request->get('search', null);
-        $users = $userRepo->getUsersSearchAutocomplete($query, 10, null, $user);
+        $users = $userRepo->getUsersSearchAutocomplete(
+            $query,
+            10,
+            null,
+            $user,
+            (!$this->isGranted(UserRoleEnum::ROLE_ADMIN)) ? $this->getSection($user) : null
```
Comment (JeremieSamson) [inline src/Controller/UserController.php]:
```suggestion
            !$this->isGranted(UserRoleEnum::ROLE_ADMIN) ? $this->getSection($user) : null
```

Comment (JeremieSamson) [reply src/Controller/UserController.php]: I don't think parenthesis are needed here

File: src/Entity/Thread.php [hunk 2]
```diff
@@ -36,11 +36,26 @@ class Thread extends BaseThread
...
      * @return ThreadMetadata[]|Collection
      */
     public function getMetadata()
     {
         return $this->metadata;
     }
+
+    public function getData(): array
+    {
+        return (null !== $this->data) ? $this->data : [];
```
Comment (JeremieSamson) [inline src/Entity/Thread.php]: same here

File: src/Form/Handler/NewThreadMessageFormHandler.php [hunk 3]
```diff
@@ -32,24 +43,71 @@ public function composeMessage(AbstractMessage $message)
...
                 ->setSubject($message->getSubject())
-                ->addRecipient($message->getRecipient())
-                ->setSender($this->getAuthenticatedParticipant())
+                ->setSender($sender)
                 ->setBody($message->getBody())
-                ->getMessage()
             ;
+
+            foreach ($participants as $participant) {
+                $thread->addRecipient($participant);
```
Comment (JeremieSamson) [inline src/Form/Handler/NewThreadMessageFormHandler.php]: Maybe a `setRecipients` would be useful here ?Will avoid to loop

File: src/Provider/MessageProvider.php [hunk 4]
```diff
@@ -3,35 +3,108 @@
...
+        $this->userRepo = $userRepo;
     }
 
     public function getThreadBySubject(string $subject): ?Thread
     {
         return $this->threadManager->findSubjectThread($subject);
     }
 
-    public function getThreadIfAlreadyExists(ParticipantInterface $recipient)
+    public function getThreadIfAlreadyExists(string $subject, array $participants)
```
Comment (JeremieSamson) [inline src/Provider/MessageProvider.php]: typehint

File: src/Provider/MessageProvider.php [hunk 5]
```diff
@@ -3,35 +3,108 @@
...
+    {
+        $sender = $this->getAuthenticatedParticipant();
+
+        return $this->threadManager->findParticipantThreadsBySubjectAndRecipients($subject, $sender, $participants);
+    }
+
+    public function getInboxThreads()
```
Comment (JeremieSamson) [inline src/Provider/MessageProvider.php]: typehint

File: src/Provider/MessageProvider.php [hunk 6]
```diff
@@ -3,35 +3,108 @@
...
+        $result = [];
+
+        foreach ($threads as $thread) {
+            array_push($result, $this->setThreadData($thread));
+        }
+
+        return $result;
+    }
+
+    public function getThread($threadId)
```
Comment (JeremieSamson) [inline src/Provider/MessageProvider.php]: typehint

File: src/Repository/UserRepository.php [hunk 7]
```diff
@@ -122,6 +124,14 @@ public function getGrantedBC(bool $includeAdmins = false): array
             $qb, $roles, true
         );
 
+        if ($section) {
+            $qb
+                ->innerJoin('u.section', 'section')
+                ->andWhere('section.id = :section')
+                ->setParameter('section', $section)
+                ;
```
Comment (JeremieSamson) [inline src/Repository/UserRepository.php]:
```suggestion
            ;
```

File: src/Controller/UserController.php (hunk 1 above)
Comment (victor-paumier) [reply src/Controller/UserController.php]: done

File: src/Entity/Thread.php (hunk 2 above)
Comment (victor-paumier) [reply src/Entity/Thread.php]: done

File: src/Provider/MessageProvider.php (hunk 4 above)
Comment (victor-paumier) [reply src/Provider/MessageProvider.php]: done

File: src/Provider/MessageProvider.php (hunk 5 above)
Comment (victor-paumier) [reply src/Provider/MessageProvider.php]: done

File: src/Provider/MessageProvider.php (hunk 6 above)
Comment (victor-paumier) [reply src/Provider/MessageProvider.php]: done

File: src/Form/Handler/NewThreadMessageFormHandler.php (hunk 3 above)
Comment (victor-paumier) [reply src/Form/Handler/NewThreadMessageFormHandler.php]: done

File: src/Form/Handler/NewThreadMessageFormHandler.php [hunk 8]
```diff
@@ -70,10 +71,7 @@ public function composeMessage(AbstractMessage $message)
                 ->setBody($message->getBody())
             ;
 
-            foreach ($participants as $participant) {
-                $thread->addRecipient($participant);
-            }
-
+            $thread->addRecipients($participants);
```
Comment (JeremieSamson) [inline src/Form/Handler/NewThreadMessageFormHandler.php]: as long as it is a new thread without any recipients before, it should be `setRecipients` don't you think ?

**Pull request merged**